src/server/frontend/wayland/wayland_connector.h

```diff
@@ -54,6 +56,30 @@ class DisplayChanger;
 class SessionAuthorizer;
 class DataDeviceManager;
 
+template<typename Callable>
+inline auto run_unless(std::shared_ptr<bool> const& condition, Callable&& callable)
+{
+    return
+        [callable = std::move(callable), condition]()
+        {
+            if (*condition)
+                return;
+
+            callable();
+        };
+}
+
+std::shared_ptr<mir::frontend::Session> session_for_client(wl_client* client);
+
+struct WlMirWindow
```

Comment:
I think that `WlMirWindow` would be better defined in wl_surface.h?

Reply:
I partly agree with @RAOF: I think wl_surface.cpp ought to work without #include "wayland_connector.h", so WlMirWindow should not be there.

However, we should be splitting WlAbstractMirWindow out too, and that has no need for the WlSurface definition. That suggests WlMirWindow would be better in wl_mir_window.h eventually and either wayland_connector.h or wl_surface.h is an acceptable intermediate step.

Reply:
I also put run_unless() in wayland_connector.h. If we want to make the Wayland .cpp files not include it, we need a better place for general purpose Wayland functions (or we move the destroyed bool and a specialized run_unless_destroyed() to the auto generated code).